### PR TITLE
Dev/threatx motion

### PR DESCRIPTION
**Pull request by kfieldho**

### CODE REVIEW

File: sprokit/processes/core/motion_detector_process.cxx [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+motion_detector_process::
+make_ports()
+{
+  // Set up for required ports
+  sprokit::process::port_flags_t required;
+  sprokit::process::port_flags_t optional;
+  sprokit::process::port_flags_t opt_static;
+
+  required.insert( flag_required );
+  opt_static.insert( flag_input_static );
```
Comment (linus-sherrill) [inline sprokit/processes/core/motion_detector_process.cxx]: Doesn't look like `opt_static` is used

**Comment by mattbrown11** [comment]
Looks good to me.

File: arrows/ocv/motion_detector_MOG2.cxx [hunk 2]
```diff
@@ -0,0 +1,281 @@
+/*ckwg +29
```
Comment (mattbrown11) [inline arrows/ocv/motion_detector_MOG2.cxx]: Are we moving away from including the copyright in each file (per a recent kitware-wide email thread)?

Comment (mleotta) [reply arrows/ocv/motion_detector_MOG2.cxx]: That has been discussed, and I would like to, but there is no action being taken as of now to make the change in KWIVER.  It requires a bit of work to update every file. Maybe this is a task for one of our software interns?

Comment (mattbrown11) [reply arrows/ocv/motion_detector_MOG2.cxx]: Can we at least decide to not care about making sure the data in each individual copyright is correct since it will be removed altogether in the near future?

Comment (mleotta) [reply arrows/ocv/motion_detector_MOG2.cxx]: I think that makes sense once we've made a firm commitment to make this change and, ideally, assigned someone to make it.  If not then the change will never happen and we'll just be in more of a mess.  @kfieldho is this something one of the software interns could tackle?

**Comment by Purg** [comment]
There is a fair about of inconsistency around parentheses spacing throughout.  Maybe we don't care for now due to possibility of using a linting/normalization process going forward.

File: arrows/ocv/three_frame_differencing.cxx [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+    //cv::cvtColor(fgmask, fgmask, CV_RGB2GRAY, 1);
+
+    if( fgmask.channels() > 1 )
+    {
+      LOG_TRACE( m_logger, "Converting multichannel foreground mask to single "
+                 "channel");
+      rms_over_channels( fgmask, fgmask, CV_8UC1);
+    }
+
+    if( true )
```
Comment (Purg) [inline arrows/ocv/three_frame_differencing.cxx]: Remove ``if( True )`` line if this is to be permanent code or attach to some verbosity parameter.

File: arrows/ocv/three_frame_differencing.cxx [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+  cv::split(src, src_split);
+  cv::Mat accum = cv::Mat(src.rows, src.cols, CV_32F, cvScalar(0));
+  for( int i=0; i<src.channels(); ++i)
+  {
+    cv::Mat temp;
+    src_split[i].convertTo(temp, CV_32F);
+
+    // Divide the result by 3^2 so that the difference has the same scale as
+    // a mono image
+    cv::multiply( temp, temp, temp, 1/9.0 );
```
Comment (Purg) [inline arrows/ocv/three_frame_differencing.cxx]: I think the 9 here should be a 3.  Duplicating these operations in equivalent numpy does not yield correct values:
```python
>>> img = np.ndarray((3, 3, 3), dtype=float)
>>> img[:] = 255
>>> # Same as current implementation
>>> np.sqrt((img * img / 9.).sum(axis=2))
array([[ 147.22431864,  147.22431864,  147.22431864],
       [ 147.22431864,  147.22431864,  147.22431864],
       [ 147.22431864,  147.22431864,  147.22431864]])
>>> # With 9 -> 3
>>> np.sqrt((img * img / 3.).sum(axis=2))
array([[ 255.,  255.,  255.],
       [ 255.,  255.,  255.],
       [ 255.,  255.,  255.]])
```

PTAL @mattbrown11

File: sprokit/processes/core/motion_detector_process.cxx [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  {
+    throw sprokit::invalid_configuration_exception( name(), "Configuration check failed." );
+  }
+
+  vital::algo::motion_detector::set_nested_algo_configuration_using_trait( algo, algo_config, d->m_algo );
+
+  if ( ! d->m_algo )
+  {
+    throw sprokit::invalid_configuration_exception( name(), "Unable to create motion detector algorithm" );
+  }
```
Comment (Purg) [inline sprokit/processes/core/motion_detector_process.cxx]: While it functionally doesn't matter given the two current ``motion_detector`` implementations, ideally the ``check_configuration`` call should also be called if ``d->m_algo`` is valid in order to allow the algorithm instance to determine if its configuration is valid for use (might not have sufficient nested algorithms, etc.).

File: sprokit/processes/core/motion_detector_process.cxx (hunk 1 above)
Comment (Purg) [reply sprokit/processes/core/motion_detector_process.cxx]: Yup, this can be removed here (probably pasted from a reference source).

File: sprokit/processes/core/motion_detector_process.cxx [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+create_config_trait( algo, std::string, "", "Algorithm configuration subblock" );
+
+//----------------------------------------------------------------
+// Private implementation class
+class motion_detector_process::priv
+{
+public:
+  priv();
+  ~priv();
```
Comment (Purg) [inline sprokit/processes/core/motion_detector_process.cxx]: Could these be given an `` = default;`` instead of empty implementations below?

File: sprokit/processes/core/register_processes.cxx [hunk 7]
```diff
@@ -138,6 +139,12 @@ register_factories( kwiver::vital::plugin_loader& vpm )
     ;
 
 
+  fact = vpm.ADD_PROCESS( kwiver::motion_detector_process );
+  fact->add_attribute( kwiver::vital::plugin_factory::PLUGIN_NAME, "motion_detector" )
+  .add_attribute( kwiver::vital::plugin_factory::PLUGIN_MODULE_NAME, module_name )
```
Comment (Purg) [inline sprokit/processes/core/register_processes.cxx]: Inconsistent spacing with regard to the rest of the file.

File: arrows/ocv/three_frame_differencing.cxx (hunk 4 above)
Comment (mattbrown11) [reply arrows/ocv/three_frame_differencing.cxx]: Yeah, should be 

`cv::multiply( temp, temp, temp, 1/3.0 );`

File: arrows/ocv/motion_detector_MOG2.h [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+  : public vital::algorithm_impl<motion_detector_MOG2,
+                                 vital::algo::motion_detector>
+{
+public:
+  double learning_rate = 0.01;
+
+  /// Constructor
+  motion_detector_MOG2();
+  /// Destructor
+  virtual ~motion_detector_MOG2() noexcept;
```
[the remark this answers is not in the document]
Comment (Purg) [reply arrows/ocv/motion_detector_MOG2.h]: Nope, not what I thought I thinking.

File: sprokit/processes/core/motion_detector_process.cxx (hunk 6 above)
Comment (Purg) [reply sprokit/processes/core/motion_detector_process.cxx]: Also not what I was thinking of.

**Comment by Purg** [comment]
Fixed most of my own comments.  Since all that's left from that list is an opinion, I will can my issues resolved.

**Comment by mattdawkins** [comment]
Minor comment - my main concern would just be naming convention, ie everything else:

  refine_detections_write_to_disk.cxx
  refine_detections_write_to_disk.cxx
  split_image.cxx
  split_image.cxx
  track_features_klt.cxx
  track_features_klt.cxx

is all verb_noun which I think we've mostly maintained across the repo

  motion_detector_3frame_differencing.cxx
  motion_detector_MOG2.cxx

also, even more minor, even though it's an abbreviation mog might look better lower case like KLT and some of the other items in folder

File: sprokit/processes/core/motion_detector_process.cxx [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+  //TODO, handle case where this is optionally provided
+  //auto ts = grab_from_port_using_trait( timestamp );
+
+  auto input = grab_from_port_using_trait( image );
+  kwiver::vital::image_container_sptr result;
+
+  {
+    scoped_step_instrumentation();
+
```
Comment (linus-sherrill) [inline sprokit/processes/core/motion_detector_process.cxx]: Usually this scope would only enclose the algorithm portion. Including the "grab_" calls here can skew results since these calls can block for an indefinite period.

**Comment by mattbrown11** [comment]
> Minor comment - my main concern would just be naming convention, ie everything else:
> 
> refine_detections_write_to_disk.cxx
> refine_detections_write_to_disk.cxx
> split_image.cxx
> split_image.cxx
> track_features_klt.cxx
> track_features_klt.cxx
> 
> is all verb_noun which I think we've mostly maintained across the repo
> 
> motion_detector_3frame_differencing.cxx
> motion_detector_MOG2.cxx
> 
I agree, we should shoot for consistency. I think the first part of the name should match the sprokit process name, in this case motion_detector. Looking through the sprokit/processes/core directory, I see a mix of nouns (video_input, matcher, keyframe_selection, image_writer) and verbs (perform_query, print_config, read_descriptor). What do you think @mleotta?

> also, even more minor, even though it's an abbreviation mog might look better lower case like KLT and some of the other items in folder

I agree.

**Comment by mleotta** [comment]
I've been trying to push for a `verb_noun` naming convention when possible, but obviously there are existing cases that don't follow this rule.  I would suggest `detect_motion_*` instead of `motion_detection_*`.

It would by nice to fix the other names that don't match the pattern at some point too, but that's a bit of work and needs to be done carefully to avoid breaking projects that use them.  For now I recommend using the pattern at least for new names so there is less to clean up later.

File: arrows/ocv/motion_detector_3frame_differencing.cxx [hunk 10]
```diff
@@ -0,0 +1,431 @@
...
+                     "Specifies the maximum expected fraction of the scene "
+                     "that may contain foreground movers at any time. When the "
+                     "fraction of pixels determined to be in motion exceeds "
+                     "this value, the background model is assumed to be "
+                     "invalid (e.g., due to excessive camera motion) and is "
+                     "reset. The default value of 1 indicates that no checking "
+                     "is done." );
+  config->set_value( "max_foreground_fract_thresh",
+                     d_->m_max_foreground_fract_thresh,
+                     "To be used in conjunction with max_foreground_fract, this"
```
[the remark this answers is not in the document]
Comment (kfieldho) [reply arrows/ocv/motion_detector_3frame_differencing.cxx]: I've added the space

File: sprokit/processes/core/motion_detector_process.cxx (hunk 9 above)
Comment (kfieldho) [reply sprokit/processes/core/motion_detector_process.cxx]: Fixed the scoping

**Comment by kfieldho** [comment]
@Purg @linus-sherrill @mattbrown11 @mattdawkins  Please take one more pass at this if you wish.  I've rebased it with the substantive changes (particular changing the naming to verb_noun) addressed, I think.  Differing the spacing issues until we decide how we'll address that repo-wide.

File: arrows/ocv/detect_motion_3frame_differencing.h [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+namespace ocv {
+
+/// OCV implementation of detect_motion using three-frame differencing
+class KWIVER_ALGO_OCV_EXPORT detect_motion_3frame_differencing
+  : public vital::algorithm_impl<detect_motion_3frame_differencing,
+                                 vital::algo::detect_motion>
+{
+public:
+  PLUGIN_INFO( "detect_motion_3frame_differencing",
+               "OCV implementation of detect_motion using three-frame differencing" )
```
Comment (linus-sherrill) [inline arrows/ocv/detect_motion_3frame_differencing.h]: The name "detect_motion_3frame_differencing" does not have to be so long. The name only has to be unique over implementations of "detect_motion" algorithms. It could be shortened to "ocv_3frame_diff".

File: arrows/ocv/detect_motion_mog2.cxx [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+namespace ocv {
+
+using namespace kwiver::vital;
+
+
+//-----------------------------------------------------------------------------
+
+// ---------------------------------------------------------------------------
+// ------------------------------ Sprokit ------------------------------------
+
```
Comment (linus-sherrill) [inline arrows/ocv/detect_motion_mog2.cxx]: It looks like these comments can be deleted.

File: arrows/ocv/detect_motion_mog2.h [hunk 13]
```diff
@@ -0,0 +1,110 @@
...
+namespace arrows {
+namespace ocv {
+
+/// OCV implementation of detect_motion_using cv::BackgroundSubtractormog2
+class KWIVER_ALGO_OCV_EXPORT detect_motion_mog2
+  : public vital::algorithm_impl<detect_motion_mog2,
+                                 vital::algo::detect_motion>
+{
+public:
+  PLUGIN_INFO( "detect_motion_mog2",
```
Comment (linus-sherrill) [inline arrows/ocv/detect_motion_mog2.h]: This name does not have to be so long. It could be shortened to "ocv_mog2"

File: sprokit/processes/core/detect_motion_process.cxx [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+// ------------------------------------------------------------------
+void
+detect_motion_process::
+_step()
+{
+  d->m_timer.start();
+
+  //TODO, handle case where this is optionally provided
+  //auto ts = grab_from_port_using_trait( timestamp );
```
Comment (linus-sherrill) [inline sprokit/processes/core/detect_motion_process.cxx]: This comment appears to have been addressed.

File: sprokit/tests/pipelines/test_motion.pipe [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+#
+# We can add a config block for two types of motion detectors
+# Only the one we select gets used
+#
+process motion
+  :: detect_motion
+
+  block algo
+    #type =  detect_motion_mog2
+    type =  detect_motion_3frame_differencing
```
Comment (linus-sherrill) [inline sprokit/tests/pipelines/test_motion.pipe]: I don't think this configuration will work as it is. The block name has to match the algorithm type string.

File: vital/algo/detect_motion.h [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+#define VITAL_ALGO_DETECT_MOTION_H
+
+
+#include <vital/vital_config.h>
+#include <vital/algo/algorithm.h>
+#include <vital/types/image_container.h>
+#include <vital/types/homography.h>
+#include <vital/types/timestamp.h>
+#include <vital/types/detected_object_set.h>
+
```
Comment (linus-sherrill) [inline vital/algo/detect_motion.h]: I think there are some extra includes here that are not needed. It looks like homography and detected_object_set are not used in the interface.

File: arrows/ocv/detect_motion_3frame_differencing.cxx [hunk 17]
```diff
@@ -0,0 +1,431 @@
...
+
+    if( fgmask.channels() > 1 )
+    {
+      LOG_TRACE( m_logger, "Converting multichannel foreground mask to single "
+                 "channel");
+      rms_over_channels( fgmask, fgmask);
+    }
+
+    {
+      double min_val, max_val;
```
[the remark this answers is not in the document]
Comment (mattbrown11) [reply arrows/ocv/detect_motion_3frame_differencing.cxx]: +1

File: arrows/ocv/detect_motion_3frame_differencing.h (hunk 11 above)
Comment (mattbrown11) [reply arrows/ocv/detect_motion_3frame_differencing.h]: +1. That makes sense since it is already and algorithm for motion detection.

File: arrows/ocv/detect_motion_mog2.cxx (hunk 12 above)
Comment (mattbrown11) [reply arrows/ocv/detect_motion_mog2.cxx]: +1

File: arrows/ocv/detect_motion_mog2.h (hunk 13 above)
Comment (mattbrown11) [reply arrows/ocv/detect_motion_mog2.h]: +1

File: sprokit/processes/core/detect_motion_process.cxx (hunk 14 above)
Comment (mattbrown11) [reply sprokit/processes/core/detect_motion_process.cxx]: +1

File: vital/algo/detect_motion.h (hunk 16 above)
Comment (mattbrown11) [reply vital/algo/detect_motion.h]: +1. I think this was a carry over from a copy-paste from other files.

File: arrows/ocv/detect_motion_3frame_differencing.cxx (hunk 17 above)
[the remark this answers is not in the document]
Comment (kfieldho) [reply arrows/ocv/detect_motion_3frame_differencing.cxx]: Fixed

File: arrows/ocv/detect_motion_3frame_differencing.h (hunk 11 above)
Comment (kfieldho) [reply arrows/ocv/detect_motion_3frame_differencing.h]: Changed to ocv_3frame_differencing

File: arrows/ocv/detect_motion_mog2.h (hunk 13 above)
Comment (kfieldho) [reply arrows/ocv/detect_motion_mog2.h]: Changed to ocv_mog2

File: sprokit/processes/core/detect_motion_process.cxx (hunk 14 above)
Comment (kfieldho) [reply sprokit/processes/core/detect_motion_process.cxx]: Comment deleted

File: sprokit/tests/pipelines/test_motion.pipe (hunk 15 above)
Comment (kfieldho) [reply sprokit/tests/pipelines/test_motion.pipe]: Fixed the pipeline names

File: vital/algo/detect_motion.h (hunk 16 above)
Comment (kfieldho) [reply vital/algo/detect_motion.h]: Removed

File: arrows/ocv/detect_motion_mog2.cxx (hunk 12 above)
Comment (kfieldho) [reply arrows/ocv/detect_motion_mog2.cxx]: removed

**Comment by Cookt2** [comment]
Jenkins test this please
